db/migrate/20220724130519_fix_post_event_timezones.rb

```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+class FixPostEventTimezones < ActiveRecord::Migration[7.0]
+  def up
+    execute <<~SQL
+      UPDATE discourse_post_event_events
+      SET
+        original_starts_at = (original_starts_at::timestamp AT TIME ZONE timezone),
+        original_ends_at = (original_ends_at::timestamp AT TIME ZONE timezone)
+      WHERE timezone IS NOT NULL;
```

Comment:
Is the `timezone` stored always UTC here? I'm a little confused here based on the description of the PR. 

> so that original_start_time and original_end_time are 'technically correct', truly-UTC timestamps.

What I understood from the above is that we'll always store the timestamps as UTC and consumers can figure out the timezone specific time by relying on the `timezone` column.

Reply:
Just to confirm my understanding, historically `original_starts_at` and `original_ends_at` have always been timestamps based on the `timezone` column and what we're trying to do here is to convert the timestamps to UTC?

Reply:
An example might be useful here. Let's say we have an event at midday in London currently (UTC +1).

Previously, we were storing it in the database as 

```
original_starts_at: 2022-07-25T12:00Z
timezone: Europe/London
```
So if you take the `original_starts_at` value in isolation, it will be one hour out. Reading the timestamp correctly would require ignoring the `Z` value and forcing `Europe/London` on the parser.

Now, we store the same event as:

```
original_starts_at: 2022-07-25T11:00Z
timezone: Europe/London
```
That means that `original_starts_at` is correct in isolation. 2022-07-025T11:00Z is technically equal to midday in London. For most consumers, that's all they need to know. If the origin timezone is required, then there are easy ruby/js APIs to translate the time into Europe/London. For example:

```
pry(main)> DateTime.parse("2022-07-25T11:00Z").in_time_zone("Europe/London")
=> Mon, 25 Jul 2022 12:00:00.000000000 BST +01:00
```

---

Followup question might be: why don't we store it as `2022-07-25T12:00+01:00` in the database? Postgres stores all timestamps internally as UTC. You can technically pass a localised timestamp to `UPDATE` or `INSERT`, but it will be immediately translated to UTC for storage.

Reply:
> Now, we store the same event as:
> 
> ```
> original_starts_at: 2022-07-25T11:00Z
> timezone: Europe/London
> ```

But based on the implementation, won't we end up storing `original_starts_at` as 13:00 instead? 

```
discourse_development=# SELECT '2022-07-25T12:00Z' AT TIME ZONE 'Europe/London';
      timezone       
---------------------
 2022-07-25 13:00:00
(1 row)
```

Reply:
> `'2022-07-25T12:00Z' AT TIME ZONE 'Europe/London';`

We strip the `Z` using `::timestamp`, so that then it is converted to 11:00:00Z

```
# SELECT '2022-07-25T12:00Z'::timestamp AT TIME ZONE 'Europe/London';
        timezone        
------------------------
 2022-07-25 11:00:00+00
(1 row)
```

This kind of hack is only used for the migration of invalid data. Going forward we won't need to do things like this.

Reply:
Ahhh OK I missed the `::timestamp` part. Thank you for clarifying :+1: